crates/bevy_render/src/mesh/mesh.rs

```diff
@@ -532,126 +531,138 @@ pub fn mesh_resource_provider_system(
         }
     }
 
+    // update changed mesh data
     for changed_mesh_handle in changed_meshes.iter() {
-        if let Some(mesh) = meshes.get(changed_mesh_handle) {
-            let vertex_bytes = mesh
-                .get_vertex_buffer_bytes(&vertex_buffer_descriptor, true)
-                .unwrap();
-            // TODO: use a staging buffer here
-            let vertex_buffer = render_resource_context.create_buffer_with_data(
-                BufferInfo {
-                    buffer_usage: BufferUsage::VERTEX,
-                    ..Default::default()
-                },
-                &vertex_bytes,
-            );
-
-            let index_bytes = mesh.get_index_buffer_bytes().unwrap();
+        if let Some(mesh) = meshes.get_mut(changed_mesh_handle) {
+            // TODO: check for individual buffer changes in non-interleaved mode
             let index_buffer = render_resource_context.create_buffer_with_data(
                 BufferInfo {
                     buffer_usage: BufferUsage::INDEX,
                     ..Default::default()
                 },
-                &index_bytes,
+                &mesh.get_index_buffer_bytes().unwrap(),
+            );
+
+            render_resource_context.set_asset_resource(
+                *changed_mesh_handle,
+                RenderResourceId::Buffer(index_buffer),
+                INDEX_BUFFER_ASSET_INDEX,
             );
 
+            // Vertex buffer
+            let vertex_count = mesh.attributes.first().unwrap().values.len(); //TODO julian: assert for different lengths of attributes, also move this somewhere else
```

Comment:
```suggestion
            let vertex_count = mesh.attributes.first().map(|a| a.values.len()).unwrap_or(0); //TODO julian: assert for different lengths of attributes, also move this somewhere else
```
This is the more panic-safe version of that basically. I had this in my own code already for doing this stuff manually :)